File: app-laravel/src/Shared/Domain/AggregateRoot.php
<?php

declare(strict_types=1);

namespace Src\Shared\Domain;

/**
 * AGGREGATE ROOT (DDD Pattern)
 * ============================
 * Aggregate Root — domen modelinin ən vacib konseptlərindən biridir.
 *
 * NƏDİR?
 * - Bir qrup əlaqəli obyekti (Entity + Value Object) birlikdə idarə edən əsas Entity-dir.
 * - Bütün dəyişikliklər YALNIZ Aggregate Root vasitəsilə edilir.
 * - Xaricdən heç vaxt daxili Entity-lərə birbaşa müraciət edilmir.
 *
 * NÜMUNƏ:
 * Order (Aggregate Root) → OrderItem (Entity) → Money (Value Object)
 * Yeni OrderItem əlavə etmək üçün birbaşa OrderItem yaratmırsan,
 * Order->addItem() metodunu çağırırsan.
 *
 * NƏYƏ LAZIMDIR?
 * - Data consistency (məlumat bütövlüyü) təmin edir.
 * - Biznes qaydalarını bir yerdə saxlayır.
 * - Domain Event-ləri burada yaranır.
 */
abstract class AggregateRoot extends Entity
{
    /**
     * Bu Aggregate daxilində baş verən domain event-lərin siyahısı.
     * Event-lər əvvəlcə burada toplanır, sonra persist edildikdən sonra dispatch olunur.
     *
     * @var DomainEvent[]
     */
    private array $domainEvents = [];

    /**
     * Yeni domain event qeydə al.
     * Bu metod yalnız Aggregate daxilindən çağırılmalıdır.
     *
     * Məsələn: Order yaradılanda $this->recordEvent(new OrderCreatedEvent(...))
     */
    protected function recordEvent(DomainEvent $event): void
    {
        $this->domainEvents[] = $event;
    }

    /**
     * Toplanmış bütün event-ləri al və siyahını təmizlə.
     * Bu metod adətən Repository-də persist-dən sonra çağırılır.
     *
     * @return DomainEvent[]
     */
    public function pullDomainEvents(): array
    {
        $events = $this->domainEvents;
        $this->domainEvents = [];

        return $events;
    }

    /**
     * Gözləyən event-lərin sayını öyrən.
     */
    public function domainEventCount(): int
    {
        return count($this->domainEvents);
    }
}

File: app/spring/src/main/java/az/ecommerce/shared/domain/AggregateRoot.java
package az.ecommerce.shared.domain;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * === AGGREGATE ROOT (DDD əsas primitivi) ===
 *
 * Laravel: src/Shared/Domain/AggregateRoot.php
 *   - protected array $domainEvents = [];
 *   - protected function recordEvent(DomainEvent $event)
 *   - public function pullDomainEvents(): array
 *
 * Spring: bu Java abstract class kimi yazılır. JPA Entity-lər bunu extend edir.
 * Domain event-lər aggregate daxilində yığılır, save-dən sonra publish olunur.
 *
 * NÜMUNƏ İSTİFADƏ:
 *   public class Order extends AggregateRoot {
 *       public void confirm() {
 *           this.status = OrderStatus.CONFIRMED;
 *           recordEvent(new OrderConfirmedEvent(this.id));
 *       }
 *   }
 */
public abstract class AggregateRoot {

    private final transient List<DomainEvent> domainEvents = new ArrayList<>();

    /**
     * Laravel: $this->recordEvent(...)
     */
    protected void recordEvent(DomainEvent event) {
        this.domainEvents.add(event);
    }

    /**
     * Laravel: $aggregate->pullDomainEvents()
     * Repository.save()-dən sonra çağrılır, sonra siyahı təmizlənir.
     */
    public List<DomainEvent> pullDomainEvents() {
        List<DomainEvent> events = Collections.unmodifiableList(new ArrayList<>(this.domainEvents));
        this.domainEvents.clear();
        return events;
    }

    public boolean hasDomainEvents() {
        return !this.domainEvents.isEmpty();
    }
}
